platform/rpc/nats/bus.go

```diff
@@ -0,0 +1,276 @@
+		if resp.Fatal {
+			return platform.FatalError{errors.New(resp.Error)}
+		}
+		return rpc.ServerError(resp.Error)
+	}
+	return nil
+}
+
+func makeErrorResponse(err error) (resp ErrorResponse) {
+	if err == nil {
+		return resp
+	}
+	if _, ok := err.(platform.FatalError); ok {
+		resp.Fatal = true
+	}
+	resp.Error = err.Error()
+	return resp
+}
+
+// requester just collect the things you need to make a request
+// together
+type requester struct {
+	conn     *nats.EncodedConn
+	instance string
+}
+
+func (r *requester) AllServices(ns string, ig flux.ServiceIDSet) ([]platform.Service, error) {
+	var response AllServicesResponse
+	if err := r.conn.Request(r.instance+methodAllServices, fluxrpc.AllServicesRequest{ns, ig}, &response, timeout); err != nil {
+		return nil, err
+	}
+	return response.Services, extractError(response.ErrorResponse)
+}
+
+func (r *requester) SomeServices(incl []flux.ServiceID) ([]platform.Service, error) {
+	var response SomeServicesResponse
+	if err := r.conn.Request(r.instance+methodSomeServices, incl, &response, timeout); err != nil {
+		return nil, err
+	}
+	return response.Services, extractError(response.ErrorResponse)
+}
+
+// Call Regrade on the remote platform. Note that we use a much longer
+// timeout, because for now at least, Regrades can take an arbitrary
+// amount of time, and we don't want to return an error if it's simply
+// taking a while. The downside is that if the platform is actually
+// not present, this won't return at all. This is somewhat mitigated
+// because regrades are done after other RPCs which have the normal
+// timeout, but better would be to split Regrades into RPCs which can
+// each have a short timeout.
+func (r *requester) Regrade(specs []platform.RegradeSpec) error {
+	var response RegradeResponse
+	if err := r.conn.Request(r.instance+methodRegrade, specs, &response, regradeTimeout); err != nil {
+		return err
+	}
+	if len(response.Result) > 0 {
+		errs := platform.RegradeError{}
+		for s, e := range response.Result {
+			errs[s] = errors.New(e)
+		}
+		return errs
+	}
+	return extractError(response.ErrorResponse)
+}
+
+func (r *requester) Ping() error {
+	var response PingResponse
+	if err := r.conn.Request(r.instance+methodPing, ping{}, &response, timeout); err != nil {
+		return err
+	}
+	return extractError(response.ErrorResponse)
+}
+
+// Connect returns a platform.Platform implementation that can be used
+// to talk to a particular instance.
+func (n *NATS) Connect(instID flux.InstanceID) (platform.Platform, error) {
+	return &requester{
+		conn:     n.snd,
+		instance: string(instID),
+	}, nil
+}
+
+// Subscribe registers a remote platform.Platform implementation as
+// the daemon for an instance (identified by instID). Any
+// platform.FatalError returned when processing requests will result
+// in the platform being deregistered, with the error put on the
+// channel `done`.
+func (n *NATS) Subscribe(instID flux.InstanceID, remote platform.Platform, done chan<- error) {
```

Comment:
is there a way of a fluxd to uniquely identify its cluster? a clusterID of some sort, it could send?